### PR TITLE
Tracker for FIPS204 / ML-DSA

### PR DESCRIPTION
**Pull request by bhess**
Tracks liboqs ML-DSA [#1919](https://github.com/open-quantum-safe/liboqs/pull/1919)

This PR Integrates ML-DSA and (non-composite) hybrids. 
However, it does not yet include support for the updated composite signatures with ML-DSA.

@feventura: It would be great if you could amend this PR to incorporate the updated composite signature support. Alternatively, you could submit a follow-up PR for those changes.

### CODE REVIEW

**Comment by dstebila** [comment]
Can we revive this PR to get ML-DSA landed in oqs-provider?  @bhess is there anything more you were planning before converting this from a draft PR?

**Comment by dstebila** [comment]
@feventura @johngray-dev do you have any feedback whether this is heaving as expected in interop with other ML-DSA implementations?

**Comment by bhess** [comment]
Tests are passing after the merge of https://github.com/open-quantum-safe/liboqs/pull/1919, so the PR is now ready.

**Comment by feventura** [comment]
> @feventura: It would be great if you could amend this PR to incorporate the updated composite signature support. Alternatively, you could submit a follow-up PR for those changes.

@bhess I look at the changes and I think it does not require any change for the composite. The composite OIDs used in there expect to use ML-DSA not dilithium.
The composite interop will still fail, but that's because I didn't implemented the -03 changes yet.

> @feventura @johngray-dev do you have any feedback whether this is heaving as expected in interop with other ML-DSA implementations?

@dstebila I don't think there is interop with this PR and other implementations, but as soon as this is merged and the [openquantumsafe/oqs-ossl3](https://hub.docker.com/r/openquantumsafe/oqs-ossl3) is updated we will see the results of the interop in the [Automated Interop Table](https://ietf-hackathon.github.io/pqc-certificates/pqc_hackathon_results_certs_r4_automated_tests.html).
Right now, oqs is failing on the ML-DSA interop, I expect that this PR will fix it.

**Review by baentsch — CHANGES_REQUESTED** [review]
Please also update the code points as per #578

File: ALGORITHMS.md
Comment (tomato42) [inline ALGORITHMS.md:75]: this is not the ID from [draft-tls-westerbaan-mldsa-00](https://datatracker.ietf.org/doc/html/draft-tls-westerbaan-mldsa-00), it should be `0x0904`

File: ALGORITHMS.md
Comment (tomato42) [inline ALGORITHMS.md:77]: given that the hybrids don't use the OIDs from [draft-ietf-lamps-pq-composite-sigs-03](https://datatracker.ietf.org/doc/html/draft-ietf-lamps-pq-composite-sigs-03), I'm assuming that those are not the LAMPS hybrids just yet?

if that's not the case, those algorithms should use IDs from [draft-reddy-tls-composite-mldsa-01](https://www.ietf.org/id/draft-reddy-tls-composite-mldsa-01.html)

File: ALGORITHMS.md
Comment (tomato42) [inline ALGORITHMS.md:83]: this is not the ID from [draft-tls-westerbaan-mldsa-00](https://datatracker.ietf.org/doc/html/draft-tls-westerbaan-mldsa-00), it should be `0x0905`

File: ALGORITHMS.md
Comment (tomato42) [inline ALGORITHMS.md:90]: this is not the ID from [draft-tls-westerbaan-mldsa-00](https://datatracker.ietf.org/doc/html/draft-tls-westerbaan-mldsa-00), it should be `0x0906`

File: ALGORITHMS.md
Comment (bhess) [reply ALGORITHMS.md]: Thanks for the pointer, @tomato42! I’ve updated the IDs based on draft-reddy-tls-composite-mldsa-01, except for the brainpool composites, which don’t seem to have an ID defined in the draft. Since I’m not fully familiar with the details of the composite implementation, @feventura, could you please confirm if these IDs align with the implementation?

File: ALGORITHMS.md
Comment (bhess) [reply ALGORITHMS.md]: updated in [2734831](https://github.com/open-quantum-safe/oqs-provider/pull/568/commits/2734831f0a25a684c969459c8198702846ccac55)

File: ALGORITHMS.md
Comment (bhess) [reply ALGORITHMS.md]: updated in [2734831](https://github.com/open-quantum-safe/oqs-provider/pull/568/commits/2734831f0a25a684c969459c8198702846ccac55)

**Comment by bhess** [comment]
Thank you, @baentsch and @tomato42, for pointing out the ML-DSA code points. I’ve updated the PR accordingly.

File: oqs-template/generate.yml
Comment (baentsch) [inline oqs-template/generate.yml:586]: Please add comment(s) pointing to the (draft) RFCs at each OID or code point where available so "posterity" can understand/track things.

Comment (bhess) [reply oqs-template/generate.yml:591]: added in [c958282](https://github.com/open-quantum-safe/oqs-provider/pull/568/commits/c9582825a049b5f957ec953b5a2c5155d8cd09f4)

File: ALGORITHMS.md
Comment (tomato42) [reply ALGORITHMS.md]: The Brainpool composites are explicitly excluded from the TLS draft, so yes, they still need to use custom codepoints.

Comment (tomato42) [reply ALGORITHMS.md]: the thing is that the lamps draft is not just concatenating two signatures: it uses the two algorithms to sign a separate synthetic message: https://datatracker.ietf.org/doc/html/draft-ietf-lamps-pq-composite-sigs-03#section-4.2.1

**Comment by SWilson4** [comment]
> @dstebila I don't think there is interop with this PR and other implementations, but as soon as this is merged and the [openquantumsafe/oqs-ossl3](https://hub.docker.com/r/openquantumsafe/oqs-ossl3) is updated we will see the results of the interop in the [Automated Interop Table](https://ietf-hackathon.github.io/pqc-certificates/pqc_hackathon_results_certs_r4_automated_tests.html). Right now, oqs is failing on the ML-DSA interop, I expect that this PR will fix it.

It would be nice to see the results of these interop tests this before we release [liboqs 0.12.0](https://github.com/orgs/open-quantum-safe/discussions/2008). @feventura If I'm understanding correctly what you wrote here, the tests will run automatically when a new version of `oqs-ossl3:latest` is pushed to Docker Hub. If that is indeed the case, what about (manually?) building and pushing a new image based on liboqs 0.12.0-rc1 and this branch of oqs-provider as part of the pre-release testing process? Tagging @baentsch and @ajbozarth as the people who likely know most about the feasibility of doing this.

**Comment by baentsch** [comment]
> what about (manually?) building and pushing a new image based on liboqs 0.12.0-rc1 and this branch of oqs-provider as part of the pre-release testing process?

Good idea and easily doable. But a bit "unfriendly" towards "unsuspecting" users of that image. @feventura : Is there any chance one can tweak the tag pulled for that automatic test? When I made these images available a few years (really?) back, I always tagged them ":ietfxyz" to ensure the right image is used at the IETF xyz hackathon event. Would you know who could change that from ":latest" to maybe also run the automated test on a tag like ":prerelease"? Or even simpler: What'd be needed for us (s part of the release process) to create such a test matrix?

**Comment by feventura** [comment]
> If I'm understanding correctly what you wrote here, the tests will run automatically when a new version of `oqs-ossl3:latest` is pushed to Docker Hub.

Not quite @SWilson4, correcting myself, the actions will not run automatically when a new version of `oqs-ossl3` is published, instead it will run when a push is done to any branch, but I or any with admin access (including you @baentsch ) could go there and trigger a new run on the git actions after the image is updated.

Now for the interop with this PR I see three options:

1. I or @baentsch  can create a branch on the hackathon repository and make change the tag for the docker image. That page would not be uploaded to the hackathon github (because the action only upload when is run from main) but the .html that would be uploaded can be downloaded.
2. Run this tests locally, and since this is already on gitaction someone could clone the hackathon repository and run the actions using [act](https://github.com/nektos/act) after making the changes on the tag for the docker image.
3. Run the test manually, here is how you do it https://github.com/IETF-Hackathon/pqc-certificates/blob/master/docs/compat_matrix_instructions.md, the only change is that we are using [R4 format](https://github.com/IETF-Hackathon/pqc-certificates?tab=readme-ov-file#zip-format-r4), not R3.

**Comment by baentsch** [comment]
> any with admin access (including you @baentsch ) 

Yikes. I still have these rights?!? I'll surely not use them without knowing exactly what I'd need to do -- I simply didn't follow the development on the hackathon github (but assumed @praveksharma does).

> Now for the interop with this PR I see three options:

Thanks for that explanation . Option 1 is imo too "heavy handed"; option 3 sounds like the fastest short-term approach and option 2 like the best "CI-able" long-term approach.

Considering @praveksharma both volunteered doing the release and also participates in the hackathon (right?) may I suggest him taking on implementing option 3 in the [release test script](https://github.com/open-quantum-safe/oqs-provider/blob/main/scripts/release-test.sh), followed by 2 in CI (or just 2 if quick-and-easy) to enhance the quality of the release testing in the long run?

**Review by baentsch — APPROVED** [review]

File: ALGORITHMS.md
Comment (tomato42) [reply ALGORITHMS.md]: I honestly don't know... if it passes interop, then yes :slightly_smiling_face:

Comment (baentsch) [reply ALGORITHMS.md]: Tagging @praveksharma : How's the IETF hackathon interop test image/run coming along?

Comment (baentsch) [reply ALGORITHMS.md]: Thanks for the confirmation via meeting that interop is OK with quite a few other implementations in the IETF hackathon. Considering the issue resolved then.

**Comment by johngray-dev** [comment]
I created some artifacts using liboqs and ML-DSA and manual testing shows they interop with the other providers in our hackathon project.   However, the docker image [openquantumsafe/oqs-ossl3](https://hub.docker.com/r/openquantumsafe/oqs-ossl3)  that is used is still failing to verify all artifacts.  Is this docker image picking up the latest 012.0 (or even the 0.12r1) version of liboqs?   I suspect it must not, otherwise the automated  [table from the hackathon](https://ietf-hackathon.github.io/pqc-certificates/pqc_hackathon_results_certs_r4_automated_tests.html)  would be showing oqs passing with positive results (instead of failing).

**Comment by baentsch** [comment]
The liboqs code base or version does not describe the functionality in an oqsprovider until both are released and documented to work together. Only an oqsprovider 0.8.0 (or as of yesterday, main) will work with liboqs 0.12.0: We had been waiting for the interop confirmation before landing the final (O)IDs. Now the docker image can follow (manually or automatically built as per preference, e.g., of @praveksharma ).

**Comment by johngray-dev** [comment]
> > @feventura: It would be great if you could amend this PR to incorporate the updated composite signature support. Alternatively, you could submit a follow-up PR for those changes.
> 
> @bhess I look at the changes and I think it does not require any change for the composite. The composite OIDs used in there expect to use ML-DSA not dilithium. The composite interop will still fail, but that's because I didn't implemented the -03 changes yet.
> 
> > @feventura @johngray-dev do you have any feedback whether this is heaving as expected in interop with other ML-DSA implementations?
> 
> @dstebila I don't think there is interop with this PR and other implementations, but as soon as this is merged and the [openquantumsafe/oqs-ossl3](https://hub.docker.com/r/openquantumsafe/oqs-ossl3) is updated we will see the results of the interop in the [Automated Interop Table](https://ietf-hackathon.github.io/pqc-certificates/pqc_hackathon_results_certs_r4_automated_tests.html). Right now, oqs is failing on the ML-DSA interop, I expect that this PR will fix it.

As mentioned above, I manually used libOQS 0.12 to interop with the other implementations in the IETF hackathon, and it works.   We are still waiting for the dockerhub image to be updated to 0.12.  Once that is done the automated interop table look much better!

**Pull request merged**
